[Misc] Add In-Container restart capability through supervisord for sagemaker entrypoint

HappyAmazonian opened the pull request:
## Purpose
- We would like to add process-level restart mechanism for Sagemaker entrypoint through our plugin https://github.com/aws/model-hosting-container-standards. The server will restart automatically within the container whenever the `vllm serve` exit accidentally.
- We provide the executable `standard-supervisor` which generates config file for `supervisor` and start the `[supervisor]`(https://supervisord.org/) process with the launch command. This enables simple integration in the vllm openai server dockerfile.
- the plugin also provides easy way for overriding the supervisor config through [environment variables](https://github.com/aws/model-hosting-container-standards/blob/main/python/model_hosting_container_standards/supervisor/README.md#configuration). 
- The feature is turned on by default. People can disable it through `PROCESS_AUTO_RECOVERY=false` env var

## Test Plan
* in our repo aws/model-hosting-container-standards, we fully test the supervisor integration  https://github.com/aws/model-hosting-container-standards/blob/main/python/tests/integration/test_supervisor_cli_integration.py

* we've also manually tested the restart by killing the vllm process inside the docker container.

To test this, we can build the sagemaker image
```
DOCKER_BUILDKIT=1 docker build . \
  --build-arg max_jobs=64 \
  --build-arg nvcc_threads=8 \
  --build-arg VLLM_USE_PRECOMPILED=1 \
  --build-arg torch_cuda_arch_list="8.0" \
  --target vllm-sagemaker \
  -t vllm/vllm-sagemaker \
  -f docker/Dockerfile

```
Start the vllm  sagemaker server 
```
docker run -it  --runtime nvidia --gpus all \
  -v ~/.cache/huggingface:/root/.cache/huggingface \
  --env "HUGGING_FACE_HUB_TOKEN=" \
  --env "SM_VLLM_MODEL=Qwen/Qwen3-30B-A3B-Thinking-2507" \
  --env "SM_VLLM_TENSOR_PARALLEL_SIZE=8" \
  --env "SM_VLLM_MAX_MODEL_LEN=4096" \
  --env "PROCESS_AUTO_RECOVERY=True" \
  --network=host \
  --ipc=host \
  vllm/vllm-sagemaker:latest
```

use ` docker exec -it [container name] bash` to log into the running container, use htop to kill the vllm subprocess and see the vllm process gets restarted

## Test Result

```
(APIServer pid=1570) INFO:     Waiting for application startup.
(APIServer pid=1570) INFO:     Application startup complete.
(APIServer pid=1570) INFO 11-11 16:03:30 [launcher.py:110] Shutting down FastAPI HTTP server.
[rank0]:[W1111 16:03:30.686624060 ProcessGroupNCCL.cpp:1524] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
(APIServer pid=1570) INFO:     Shutting down
(APIServer pid=1570) INFO:     Waiting for application shutdown.
(APIServer pid=1570) INFO:     Application shutdown complete.
# Supervisor detected the process exited
2025-11-11 16:03:33,577 WARN exited: app (exit status 0; not expected)
2025-11-11 16:03:34,579 INFO spawned: 'app' with pid 1955
2025-11-11 16:03:35,581 INFO success: app entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
## Restart the vllm serve
(APIServer pid=1955) INFO 11-11 16:03:43 [api_server.py:1897] vLLM API server version 0.11.1rc5.dev367+g40cd1ed24
(APIServer pid=1955) INFO 11-11 16:03:43 [utils.py:253] non-default args: {'model_tag': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0', 'model': 'TinyLlama/TinyLlama-1.1B-Chat-v1.0'}
(APIServer pid=1955) INFO 11-11 16:03:44 [model.py:630] Resolved architecture: LlamaForCausalLM
(APIServer pid=1955) INFO 11-11 16:03:44 [model.py:1728] Using max model len 2048
(APIServer pid=1955) INFO 11-11 16:03:44 [scheduler.py:254] Chunked prefill is enabled with max_num_batched_tokens=2048.
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 0c028ba489f560102a87a3e98b92caac5a6da8f1. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Wraps `vllm serve` with `standard-supervisor` in `sagemaker-entrypoint.sh` to enable in-container process auto-restart and supervisor-based configuration via environment variables.
> 
> - Replaces direct `exec vllm serve ...` with `exec standard-supervisor vllm serve ...`
> - Keeps existing env-to-CLI arg translation and required port setup unchanged
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0c028ba489f560102a87a3e98b92caac5a6da8f1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervisor-based auto-restart for the SageMaker example server.
> 
> - Replaces direct `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `sagemaker-entrypoint.sh` to run under `supervisord`
> - Bumps `model-hosting-container-standards` to `>= 0.1.13, < 1.0.0` in `requirements/common.txt` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 798f358dab6913beb4138f2fcde7d460d89e8c1f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervisor-based auto-recovery for the SageMaker example server.
> 
> - Wraps `vllm serve` with `standard-supervisor` in `examples/online_serving/sagemaker-entrypoint.sh` to run under `supervisord` for in-container restart and env-driven config
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a41c7af3bfe5f39bb132d2e4d06a59d22d804a18. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervisor-based auto-restart for the SageMaker example server.
> 
> - Replaces `exec vllm serve` with `exec standard-supervisor vllm serve` in `examples/online_serving/sagemaker-entrypoint.sh` to run under `supervisord`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 13f4ad0f55b08ca1d37d1459f50ce950feaeb547. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised execution of the SageMaker example server for in-container process restart and env-based supervisor configuration.
> 
> - Wraps `vllm serve` with `standard-supervisor` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ca8f8e90fffd35fe29666bb30f3a19fa198f53b0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised execution and auto-restart for the SageMaker example server.
> 
> - Replaces `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3022f4b6b7b174dbdabca7115cb9135a598a3a15. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised execution and in-container auto-restart for the SageMaker example server.
> 
> - Replaces `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3022f4b6b7b174dbdabca7115cb9135a598a3a15. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces supervised execution for the SageMaker example to enable in-container auto-restart and env-driven supervision.
> 
> - Replaces direct `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2909df18c3b08526c583e0830376b46db73e045. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised, auto-restarting execution for the SageMaker example server.
> 
> - Wraps `vllm serve` with `standard-supervisor` in `examples/online_serving/sagemaker-entrypoint.sh` to run under `supervisord`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2909df18c3b08526c583e0830376b46db73e045. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised execution and auto-recovery for the SageMaker example server.
> 
> - Replaces `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh` to run under `supervisord` (env→CLI arg translation remains unchanged)
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0` to support the supervisor wrapper
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2909df18c3b08526c583e0830376b46db73e045. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables supervised execution and in-container auto-restart for the SageMaker example server.
> 
> - Replaces direct `exec vllm serve` with `exec standard-supervisor vllm serve` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bumps `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b2909df18c3b08526c583e0830376b46db73e045. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds supervised execution for in-container auto-restart and env-driven config.
> 
> - Replace `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bump `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2a02d0083fd25fe888c0c894bccc3f09e4e6874b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces supervised execution for the SageMaker example to enable in-container auto-restart.
> 
> - Replace `exec vllm serve ...` with `exec standard-supervisor vllm serve ...` in `examples/online_serving/sagemaker-entrypoint.sh`
> - Bump `model-hosting-container-standards` in `requirements/common.txt` to `>= 0.1.13, < 1.0.0`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit cff190a3d362e26cf5915ca27a203612ef08bd44. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

Lokiiiiii commented:
> Question, typically our users rely on the cluster orchestrator to perform the restart, such as
> 
>     * Docker's restart flag
> 
>     * Kubernetes will redeploy the pod on crash
> 
> 
> This also gives granular observabilities into statistics of how often the pod crashes and restarts
> 
> Can you please leave some insights?

These are valid points, we can disable this feature by default but consider this as a first line of defense that is useful in the following scenarios - 
1. Legacy managed deployment systems that do not use container orchestrators
1. Reducing recovery time by preserving local artifacts like Compilation cache, HF cache, etc.
1. Reducing recovery time by avoiding new container file system creation

We are trying to bring process resiliency without being opinionated on deployment patterns.

HappyAmazonian committed:
[Misc] Add In-Container restart capability through supervisord for openai server #28502

Signed-off-by: Shen Teng <sheteng@amazon.com>

HappyAmazonian commented:
> Since it looks like this does not require any new dependencies, I'm interested to see what an "off by deafult" approach would look like. If it's simple enough, it seems OK. Otherwise, I'd say I think this sounds like custom behavior that should be enabled by building your own image. Let's see what having it as an option looks like, though.

Hi. Thanks for replying. We changed our plugin. So now the by default, it uses `os.execvp` to pass the command. No behavioral change to the existing code. People can enable auto-restart by setting PROCESS_AUTO_RECOVERY=1

reference: https://github.com/aws/model-hosting-container-standards/blob/main/python/model_hosting_container_standards/supervisor/scripts/standard_supervisor.py#L199

simon-mo commented:
Is `standard-supervisor` from Sagemaker? This makes our container default depends on a third party library for entrypoint, which is a bit risky.

HappyAmazonian committed:
feat(docker): make standard-supervisor optional via entrypoint script

Add configurable entrypoint.sh that allows toggling standard-supervisor
usage via VLLM_USE_STANDARD_SUPERVISOR environment variable.

- Default: direct vllm serve execution (standard-supervisor disabled)
- Set VLLM_USE_STANDARD_SUPERVISOR=true to enable standard-supervisor

Signed-off-by: Shen Teng <sheteng@amazon.com>

simon-mo reviewed (changes requested):
Hi @HappyAmazonian, thank you for the effort on this. 

There's already a sagemaker container tag (see few lines below) that users can trivially build. Please don't change the default entrypoint for vllm image, we would like to it to remain `vllm serve`

chaunceyjiang reviewed (commented):
Whether you use Docker, contaienrd. Podman, or Kubernetes, you can override the `ENTRYPOINT` at startup.

Why do we need to modify the Dockerfile then?

HappyAmazonian committed:
Remove docker entrypoint wrapper and use vllm serve directly

- Remove docker/entrypoint.sh wrapper script
- Update Dockerfile to use 'vllm serve' as direct entrypoint
- Update sagemaker-entrypoint.sh to use standard-supervisor

Signed-off-by: Shen Teng <sheteng@amazon.com>

HappyAmazonian committed:
Merge branch 'main' into restart

Signed-off-by: HappyAmazonian <91216626+HappyAmazonian@users.noreply.github.com>

HappyAmazonian committed:
Merge branch 'vllm-project:main' into restart

HappyAmazonian committed:
Update common.txt requirements

Signed-off-by: Shen Teng <sheteng@amazon.com>

HappyAmazonian committed:
Merge branch 'restart' of https://github.com/HappyAmazonian/vllm into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart

HappyAmazonian committed:
Merge branch 'main' into restart
